File: 1. JPA basic/basic/src/main/java/study/jpa/basic/jpashop/domain/Category.java
package study.jpa.basic.jpashop.domain;

import java.util.ArrayList;
import java.util.List;

import javax.persistence.*;

/**
 * @author Joonhyuck Hyoung
 */
@Entity(name = "jpashop_category")
@Table(name = "jpashop_category")
public class Category extends BaseEntity {
    @Id
    @GeneratedValue
    private Long no;

    private String name;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "PARENT_ID")
    private Category parent;

    @OneToMany(mappedBy = "parent") // 여긴 기본이 LAZY
    private List<Category> child = new ArrayList<>();

    // item과 다대다
    @ManyToMany
    @JoinTable(name = "CATEGORY_ITEM",
            joinColumns = @JoinColumn(name = "CATEGORY_ID"),
            inverseJoinColumns = @JoinColumn(name = "ITEM_ID")
    )
    private List<Item> items = new ArrayList<>();
}

File: 2. JPA deep/jpashop/src/main/kotlin/jpabook/jpashop/domain/Category.kt
package jpabook.jpashop.domain

import jpabook.jpashop.domain.item.*
import javax.persistence.*

@Entity
class Category(
    @Id @GeneratedValue
    @Column(name = "category_id")
    val id: Long,

    val name: String,

    /**
     * @ManyToMany
     * category_item에 컬럼을 추가할 수 없고, 세밀한 쿼리를 실행하기 어려우니까 실무에서 사용하지 말자
     * 중간 엔티티(CategoryItem)를 만들고 @ManyToOne, @OneToMany로 매핑해서 사용하자
     */
    @ManyToMany
    @JoinTable(
        name = "category_item",
        joinColumns = [JoinColumn(name = "category_id")],
        inverseJoinColumns = [JoinColumn(name = "item_id")]
    )
    val items: MutableList<Item> = arrayListOf(),

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    var parent: Category,

    @OneToMany(mappedBy = "parent")
    val child: MutableList<Category> = arrayListOf(),
) {
    //==연관관계 메서드==//
    fun addChildCategory(child: Category) {
        this.child.add(child)
        child.parent = this
    }
}
